### ISSUE TITLE
[expo-notifications][ios] Error when using removeAllNotificationListeners with fast refresh

### ISSUE THREAD

**Issue by brunoaduarte** [expo/expo#9106]
### Summary of Issue 

> Attempted to remove more 'ExpoNotificationsEmitter' listeners than added.
Error running the basic expo-notifications example.
Caused by this line
`Notifications.removeAllNotificationListeners();`

### Environment - output of `expo diagnostics` & the platform(s) you're targeting

>   Expo CLI 3.21.13 environment info:
>     System:
>       OS: Windows 10 10.0.18362
>     Binaries:
>       Node: 14.4.0 - C:\Program Files\nodejs\node.EXE
>       npm: 6.14.4 - C:\Program Files\nodejs\npm.CMD
>     npmPackages:
>       expo: ~38.0.8 => 38.0.8
>       react: ~16.11.0 => 16.11.0
>       react-dom: ~16.11.0 => 16.11.0
>       react-native: https://github.com/expo/react-native/archive/sdk-38.0.1.tar.gz => 0.62.2
>       react-native-web: ~0.11.7 => 0.11.7

Target: iOS (iPhone 6 real device)

### Reproducible Demo

```
import Constants from 'expo-constants';
import * as Notifications from 'expo-notifications';
import * as Permissions from 'expo-permissions';
import React, { useState, useEffect } from 'react';
import { Text, View, Button, Platform } from 'react-native';

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false,
  }),
});

export default function App() {
  const [expoPushToken, setExpoPushToken] = useState('');
  const [notification, setNotification] = useState(false);

  useEffect(() => {
    registerForPushNotificationsAsync().then(token => setExpoPushToken(token));
    Notifications.addNotificationReceivedListener(notification => {
      setNotification(notification);
    });

    return () => {
      Notifications.removeAllNotificationListeners();
    };
  });

  return (
    <View
      style={{
        flex: 1,
        alignItems: 'center',
        justifyContent: 'space-around',
      }}>
      <Text>Your expo push token: {expoPushToken}</Text>
      <View style={{ alignItems: 'center', justifyContent: 'center' }}>
        <Text>Title: {notification && notification.request.content.title} </Text>
        <Text>Body: {notification && notification.request.content.body}</Text>
        <Text>Data: {notification && JSON.stringify(notification.request.content.data.body)}</Text>
      </View>
      <Button
        title="Press to Send Notification"
        onPress={async () => {
          await sendPushNotification(expoPushToken);
        }}
      />
    </View>
  );
}

// Can use this function below, OR use Expo's Push Notification Tool-> https://expo.io/dashboard/notifications
async function sendPushNotification(expoPushToken) {
  const message = {
    to: expoPushToken,
    sound: 'default',
    title: 'Original Title',
    body: 'And here is the body!',
    data: { data: 'goes here' },
  };

  await fetch('https://exp.host/--/api/v2/push/send', {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Accept-encoding': 'gzip, deflate',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(message),
  });
}

async function registerForPushNotificationsAsync() {
  let token;
  if (Constants.isDevice) {
    const { status: existingStatus } = await Permissions.getAsync(Permissions.NOTIFICATIONS);
    let finalStatus = existingStatus;
    if (existingStatus !== 'granted') {
      const { status } = await Permissions.askAsync(Permissions.NOTIFICATIONS);
      finalStatus = status;
    }
    if (finalStatus !== 'granted') {
      alert('Failed to get push token for push notification!');
      return;
    }
    token = (await Notifications.getExpoPushTokenAsync()).data;
    console.log(token);
  } else {
    alert('Must use physical device for Push Notifications');
  }

  if (Platform.OS === 'android') {
    Notifications.setNotificationChannelAsync('default', {
      name: 'default',
      importance: Notifications.AndroidImportance.MAX,
      vibrationPattern: [0, 250, 250, 250],
      lightColor: '#FF231F7C',
    });
  }

  return token;
}

```
 
### Steps to Reproduce

```
expo init bugTest (select blank)
cd bugTest
expo install expo-notifications
expo start
```

### Expected Behavior vs Actual Behavior

![image](https://user-images.githubusercontent.com/5354345/86620595-f3860b00-bf92-11ea-9c42-369006ff790a.png)

**Comment by brunoaduarte** [comment]
Workaround

```
...

useEffect(() => {
    registerForPushNotificationsAsync().then(token => setExpoPushToken(token));

    const notificationSubscription = Notifications.addNotificationReceivedListener(notification => {
      setNotification(notification);
    });

    return () => {
      Notifications.removeNotificationSubscription(notificationSubscription);
    };

  });

...
```

**Comment by lukmccall** [comment]
I believe that is a bug in this demo. This should look like this:
```ts
  useEffect(() => {
    registerForPushNotificationsAsync().then(token => {
      console.log(token);
      setExpoPushToken(token);
    });

    Notifications.addNotificationReceivedListener(notification => {
      console.log(notification);
      setNotification(notification);
    });

    Notifications.addNotificationResponseReceivedListener(
      response => {
        console.log(response);
      }
    );

    return () => {
      Notifications.removeAllNotificationListeners();
    };
  }, []); // <- see here
```

**Comment by pascalchristian** [comment]
> I believe that is a bug in this demo. This should look like this:
> 
> ```ts
>   useEffect(() => {
>     registerForPushNotificationsAsync().then(token => {
>       console.log(token);
>       setExpoPushToken(token);
>     });
> 
>     Notifications.addNotificationReceivedListener(notification => {
>       console.log(notification);
>       setNotification(notification);
>     });
> 
>     Notifications.addNotificationResponseReceivedListener(
>       response => {
>         console.log(response);
>       }
>     );
> 
>     return () => {
>       Notifications.removeAllNotificationListeners();
>     };
>   }, []); // <- see here
> ```

I have this code on my live project (use effect run only once), but still stumbles into the same bug in ios only. Specifically with live reload enabled on the expo mobile client.

**Comment by cruzach** [comment]
@pascalchristian I updated the demo, but can confirm that I see the same behavior on iOS when running the snack locally with expo-cli (even with the updated code)

When I switch to removing the subscriptions one by one with `Notifications.removeNotificationSubscription(subscription)`, like this:
```
useEffect(() => {
  // add sub1 and sub2
...
  return () => {
      Notifications.removeNotificationSubscription(sub1);
      Notifications.removeNotificationSubscription(sub2);
    };
  }, []);
```

I no longer see this error developing locally

**Comment by lukmccall** [comment]
Ok then, it looks like the bug is in the `UMReactNativeEventEmitter` - see `https://github.com/expo/expo/blob/62aa41504c39c964c3ba0fdffc82b69f1676e047/packages/%40unimodules/react-native-adapter/ios/UMReactNativeAdapter/Services/UMReactNativeEventEmitter.m#L87-L124`. 
For some reasons, this method is called tree times 😕

**Comment by mobeendev** [comment]
Any update on this?

**Comment by sjchmiela** [comment]
Yes! The fixing pull request is awaiting review. Sorry it took so long.

**Comment by awilliamsra** [comment]
I have consumed the latest Expo 39 SDK version (39.0.3) - but this still seems to throw the same error.

**Comment by lukmccall** [comment]
@awilliamsra please open a new issue ;)
